### PR TITLE
Grid extension to return mulitiple points located in given search radius

### PR DESCRIPTION
**Pull request by norihiro-w**
reworked on #789. shall be rebased after #795.

### CODE REVIEW

File: GeoLib/Grid.h [hunk 1]
```diff
@@ -325,7 +120,7 @@ class Grid : public GeoLib::AABB<POINT>
 	 * @param coords (output) the coordinates of the grid cell
 	 */
 	template <typename T>
-	inline void getGridCoords(T const& pnt, std::size_t* coords) const;
+	inline std::array<std::size_t,3> getGridCoords(T const& pnt) const;
```
Comment (endJunction) [inline GeoLib/Grid.h]: I think `inline` is not necessary here, isn't it?

File: GeoLib/Grid.h [hunk 2]
```diff
@@ -384,29 +376,42 @@ void Grid<POINT>::getGridCoords(T const& pnt, std::size_t* coords) const
...
-						&& dists[2] >= min_dist
-						&& dists[3] >= min_dist
-						&& dists[4] >= min_dist
-						&& dists[5] >= min_dist) {
+		if (dists[0] >= min_dist && dists[1] >= min_dist
+			&& dists[2] >= min_dist && dists[3] >= min_dist
+			&& dists[4] >= min_dist && dists[5] >= min_dist) {
 			return nearest_pnt;
 		}
 	} else {
```
Comment (endJunction) [inline GeoLib/Grid.h]: The else branch seems not to be covered by the tests (and was not before). So this is just a remark.
@TomFischer can you extend the tests, such the else branch is taken in the feature?

File: GeoLib/Grid.h [hunk 3]
```diff
@@ -50,7 +51,7 @@ class Grid : public GeoLib::AABB<POINT>
 	 *
 	 */
 	template <typename InputIterator>
-	Grid(InputIterator first, InputIterator last, std::size_t max_num_per_grid_cell = 512);
+	Grid(InputIterator first, InputIterator last, std::size_t items_per_cell = 512);
```
Comment (endJunction) [inline GeoLib/Grid.h]: Please update documentation above for all parameters.

File: GeoLib/Grid.h [hunk 4]
```diff
@@ -88,8 +91,8 @@ class Grid : public GeoLib::AABB<POINT>
 	 * the axis aligned cube
 	 */
 	template <typename P>
-	void getPntVecsOfGridCellsIntersectingCube(P const& center, double half_len,
-		std::vector<std::vector<POINT*> const*>& pnts) const;
+	std::vector<std::vector<POINT*> const*>
+	getPntVecsOfGridCellsIntersectingCube(P const& center, double half_len) const;
```
Comment (endJunction) [inline GeoLib/Grid.h]: Update docu; `pnts` to \return

File: GeoLib/Grid.h (hunk 1 above)
Comment (endJunction) [inline GeoLib/Grid.h]: Remove coords from docu; Return value.

File: GeoLib/Grid.h [hunk 5]
```diff
@@ -150,13 +153,12 @@ class Grid : public GeoLib::AABB<POINT>
 	 * @param coords coordinates of the grid cell
 	 */
 	template <typename P>
-	void getPointCellBorderDistances(P const& pnt,
-	                                 double dists[6],
-	                                 std::size_t const* const coords) const;
+	std::array<double, 6> getPointCellBorderDistances(
+		P const& pnt, std::array<std::size_t,3> const& coords) const;
```
Comment (endJunction) [inline GeoLib/Grid.h]: Docu on parameters.

**Comment by endJunction** [comment]
:fast_forward: after doxygen docu updates.

File: Tests/MeshLib/TestMeshNodeSearch.cpp [hunk 6]
```diff
@@ -60,6 +61,64 @@ class MeshLibMeshNodeSearchInSimpleHexMesh : public testing::Test
...
+	}
+	{
+		GeoLib::Point p(dx_half*1.01, 0.0, 0.0);
+		ASSERT_EQ(1u, mesh_node_searcher.getMeshNodeIDsForPoint(p).size());
+		ASSERT_EQ(1u, mesh_node_searcher.getMeshNodeIDsForPoint(p)[0]);
+	}
+}
+
+/*
+TEST_F(MeshLibMeshNodeSearchInSimpleQuadMesh, PointSearch3)
```
Comment (endJunction) [inline Tests/MeshLib/TestMeshNodeSearch.cpp]: Forgot to mention this one. Either add or remove or at least comment why commented.

Comment (norihiro-w) [inline Tests/MeshLib/TestMeshNodeSearch.cpp]: removed

**Comment by norihiro-w** [comment]
I added `MeshNodesOnPoint` class which caches a result of `Grid::getPointsInEpsilonEnvironment()`

**Comment by TomFischer** [comment]
Opened PR https://github.com/ufz/ogs/pull/810 where I rebased the PR and reworked class Grid a little bit more.

**Comment by endJunction** [comment]
I think we can move to ufz/ogs#810.

**Comment by norihiro-w** [comment]
@TomFischer thank you for continuing working on the Grid class

**Pull request closed without merging**